contributing/topics/guide-state-migrations.md

```diff
@@ -0,0 +1,170 @@
+		return nil, nil
+	}
+}
+```
+
+3. Copy over the schema for `capybara_resource.go`. If nothing in the schema is changing then this can be copied over 1:1, however you will want to go through and remove some property attributes that are not required. 
+The information in `Schema()` is used to help core serialize/deserialize the values when sending the data over RPC. For this reason the following property attributes should be removed from the schema since they unnecessarily bloat the code base:
+   * Default
+   * ValidateFunc
+   * ForceNew
+   * MaxItems
+   * MinItems
+   * AtLeastOneOf
+   * ConflictsWith
+   * ExactlyOneOf
+   * RequiredWith
+4. Fill out the UpgradeFunc to make the modification to the Resource ID. In most cases this involves parsing the old ID insensitively and then overwriting the value for `id` in the state. The file should now look like this:
+```go
+package migration
+
+import (
+	"context"
+	"log"
+
+	"github.com/hashicorp/go-azure-sdk/resource-manager/animals/2023-11-01/capybaras"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+)
+
+type CapybaraV0ToV1 struct{}
+
+func (s CapybaraV0ToV1) Schema() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"name": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+		},
+
+		"cuteness": {
+			Type:     pluginsdk.TypeInt,
+			Required: true,
+		},
+
+		"pet_names": {
+			Type:     pluginsdk.TypeList,
+			Optional: true,
+			Elem: &pluginsdk.Schema{
+				Type: pluginsdk.TypeString,
+			},
+		},
+	}
+}
+
+func (s CapybaraV0ToV1) UpgradeFunc() pluginsdk.StateUpgraderFunc {
+	return func(ctx context.Context, rawState map[string]interface{}, meta interface{}) (map[string]interface{}, error) {
+		oldId := rawState["id"].(string)
+		newId, err := capybaras.ParseCapybaraIDInsensitively(oldId)
+		if err != nil {
+			return nil, err
+		}
+
+		log.Printf("[DEBUG] Updating ID from %q to %q", oldId, newId)
+
+		rawState["id"] = newId.ID()
+		return rawState, nil
+	}
+}
+```
+
+5. Finally we hook the state migration up to the resource. For typed resources this looks like the following
+```go
+package animal
+
+import (
+	"context"
+	"fmt"
+	"time"
+	
+	"github.com/hashicorp/go-azure-sdk/resource-manager/animals/2023-11-01/capybaras"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/sdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/animals/migration"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+)
+
+type CapybaraResource struct{}
+
+var (
+	_ sdk.ResourceWithStateMigration = CapybaraResource{}
+)
+
+type CapybaraResourceModel struct {
+	Name       string   `tfschema:"name"`
+	Cuteness   string   `tfschema:"cuteness"`
+	PetNames   []string `tfschema:"pet_names"`
+
+}
+
+func (r CapybaraResource) StateUpgraders() sdk.StateUpgradeData {
+	return sdk.StateUpgradeData{
+		SchemaVersion: 1,
+		Upgraders: map[int]pluginsdk.StateUpgrade{
+			0: migration.CapybaraV0ToV1{},
+		},
```

Comment:
We should add a comment explaining why `StateVersion` is set to `1` here (because it's the next version) - whilst it's present in the docs for this field, it's not necessarily clear from this example?